### PR TITLE
fix(json-abi): properly handle Param `type` field

### PR DESCRIPTION
**Pull request by DaniPopes**
## Motivation

`type` is either:
- "tuple" + [] array brackets if `components` is not empty
- a canonical Solidity type if `components` is empty

Check this during deserialization and signature formatting.

Added benchmark for signature formatting:

```
signature/large-function/alloy                                                                            
                        time:   [217.91 ns 218.47 ns 219.13 ns]
signature/large-function/ethabi                                                                             
                        time:   [2.6704 µs 2.6737 µs 2.6768 µs]
```

## PR Checklist

- [x] Added Tests
- [x] Added Documentation
- [ ] Breaking changes

### CODE REVIEW

File: crates/json-abi/src/param.rs [hunk 1]
```diff
@@ -184,6 +180,12 @@ impl Param {
             s.push_str(&self.ty)
         } else {
             crate::utils::signature_raw("", &self.components, s);
+            // checked during deserialization, but might be invalid from a user
+            debug_assert!(
+                self.ty.starts_with("tuple"),
+                "expected tuple param type to start with \"tuple\"",
+            );
+            s.push_str(unsafe { self.ty.get_unchecked(5..) });
```
Comment (DaniPopes) [inline crates/json-abi/src/param.rs:188]: Should we allow this with `if let Some = strip_prefix` ?

File: crates/json-abi/src/param.rs [hunk 2]
```diff
@@ -184,6 +180,12 @@ impl Param {
             s.push_str(&self.ty)
         } else {
             crate::utils::signature_raw("", &self.components, s);
+            // checked during deserialization, but might be invalid from a user
```
Comment (prestwich) [inline crates/json-abi/src/param.rs:183]: could we check this on type-instantiation so it's impossible for a user to construct an invalid one?

Comment (DaniPopes) [reply crates/json-abi/src/param.rs:183]: How? All struct fields are public

Comment (prestwich) [reply crates/json-abi/src/param.rs:183]: if public fields allow the user to create invalid instances, then we should at least consider making them non-public

**Review by prestwich — APPROVED** [review]

**Pull request merged**